WorkerNavigator::gpu() can only be invoked from a worker

annevk opened the pull request:
#### 4635a5d12b4bcaf0c05d88c9e22ee9f408f09ac6
<pre>
WorkerNavigator::gpu() can only be invoked from a worker
<a href="https://bugs.webkit.org/show_bug.cgi?id=293800">https://bugs.webkit.org/show_bug.cgi?id=293800</a>

Reviewed by Mike Wyrzykowski.

* Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
(WebCore::NavigatorGamepad::protectedNavigator const): Deleted.

This method is unneeded as we can const m_navigator instead.

* Source/WebCore/page/Navigator.cpp:

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::gpu):

As this method can only be invoked from a worker, it does not make
sense for it to implement logic for documents.

(WebCore::WorkerNavigator::setAppBadge):

Adopt dynamicDowncast&lt;&gt; and RefPtr.

Canonical link: <a href="https://commits.webkit.org/295642@main">https://commits.webkit.org/295642@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78a7f695f5b120dd9c5db2fe9355b7446b34c87b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105647 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25396 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15793 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110840 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56242 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107688 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25868 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33903 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80236 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108653 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20232 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95350 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60542 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19968 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13439 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55678 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89681 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13482 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113688 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32785 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24181 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89316 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33147 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91573 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88974 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33848 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11657 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28242 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17155 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32710 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38105 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32456 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35805 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34054 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/page/WorkerNavigator.cpp:
```diff
@@ -65,28 +65,15 @@ GPU* WorkerNavigator::gpu()
-
-            m_gpuForWebGPU = GPU::create(*gpu);
-        }
+        Ref context = downcast<WorkerGlobalScope>(*this->scriptExecutionContext());
```
Does this work for DedicatedWorkers too? It looks like the:
```
} else if (scriptExecutionContext->isDocument()) {
            Ref document = downcast<Document>(*scriptExecutionContext);
```
was added for dedicated workers, but maybe it is not needed?

annevk replied:
Dedicated workers also use `WorkerGlobalScope`. They're not documents...

mwyrzykowski replied:
Sorry I just checked, indeed none of the workers take the removed path.

mwyrzykowski reviewed (approved):
Thank you!

annevk committed:
WorkerNavigator::gpu() can only be invoked from a worker
https://bugs.webkit.org/show_bug.cgi?id=293800

Reviewed by Mike Wyrzykowski.

* Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
(WebCore::NavigatorGamepad::protectedNavigator const): Deleted.

This method is unneeded as we can const m_navigator instead.

* Source/WebCore/page/Navigator.cpp:

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::gpu):

As this method can only be invoked from a worker, it does not make
sense for it to implement logic for documents.

(WebCore::WorkerNavigator::setAppBadge):

Adopt dynamicDowncast<> and RefPtr.

Canonical link: https://commits.webkit.org/295642@main
